akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala

```diff
@@ -448,31 +448,36 @@ private[cluster] class ClusterCoreDaemon(publisher: ActorRef) extends Actor with
 
       // check by address without uid to make sure that node with same host:port is not allowed
       // to join until previous node with that host:port has been removed from the cluster
-      val alreadyMember = localMembers.exists(_.address == node.address)
-      val isUnreachable = !latestGossip.overview.reachability.isReachable(node)
+      localMembers.find(_.address == node.address) match {
+        case Some(m) if m.uniqueAddress == node ⇒
+          // node retried join attempt, probably due to lost Welcome message
+          logInfo("Existing member [{}] is joining again.", node)
+          if (node != selfUniqueAddress)
+            sender() ! Welcome(selfUniqueAddress, latestGossip)
+        case Some(m) ⇒
+          // node restarted, same host:port as existing member, but with different uid
+          // safe to down and later remove existing member
+          // new node will retry join
+          logInfo("New incarnation of existing member [{}] is trying to join. " +
+            "Existing will be removed from the cluster and then new member will be allowed to join.", node)
+          if (m.status != Down && m.status != Leaving && m.status != Exiting)
+            downing(m.address)
```

Comment:
no, the identifier of a member is hostname+port+uid, that will not change with the support of binding to a different external address, unless I have missed something completely (and then we have major problems with that host binding feature).

If we get a join request with new uid and same hostname+port as existing member it is know that the old uid is gone, and can be downed (independent of status or reachability of existing member). It will become unreachable if it's not already unreachable and then it can be removed by the normal removal procedure.